src/framework/preferences/js/SelfVoicingEnactor.js

```diff
@@ -92,45 +108,280 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
+                        } else if (word && fluid.prefs.enactor.selfVoicing.isWord(fluid.get(parsed, [(parsed.length - 1), "word"]))) {
+                            parsed.push({
+                                blockIndex: blockIndex,
+                                startOffset: charIndex,
+                                endOffset: charIndex + word.length,
+                                node: childNode,
+                                childIndex: childIndex,
+                                parentNode: childNode.parentNode,
+                                word: word
+                            });
+                            blockIndex += word.length;
+                        }
+                        charIndex += word.length;
+                    });
+                } else if (childNode.nodeType === fluid.prefs.enactor.selfVoicing.nodeType.ELEMENT_NODE && fluid.prefs.enactor.selfVoicing.hasRenderedText(childNode)) {
+                    parsed = parsed.concat(fluid.prefs.enactor.selfVoicing.parse(childNode, blockIndex));
+                    if (parsed.length) {
+                        var lastParsed = parsed[parsed.length - 1];
+                        blockIndex = lastParsed.blockIndex + lastParsed.word.length;
                     }
-                } else {
-                    fluid.prefs.enactor.selfVoicing.readFromDOM(that, node);
                 }
-            }
+            });
+        }
+
+        return parsed;
+    };
+
+    /**
+     * Combines the parsed text into a String.
+     *
+     * @param {Array} parsed - An array of parsed data points
+     *
+     * @return {String} - The parsed text combined into a String.
+     */
+    fluid.prefs.enactor.selfVoicing.parsedToString = function (parsed) {
+        var words = fluid.transform(parsed, function (block) {
+            return block.word;
         });
+
+        return words.join("");
+    };
+
+    /**
+     * Parses the DOM element into data points to use for highlighting the text, and queues the text into the self
+     * voicing engine. The parsed data points are added as an array to the component's `parseQueue`
+     *
+     * @param {Component} that - the component
+     * @param {node} elm - The DOM node to read
+     */
+    fluid.prefs.enactor.selfVoicing.readFromDOM = function (that, elm) {
+        elm = $(elm);
+
+        // only execute if there are nodes to read from
+        if (elm.length) {
+            var parsedFromElm = fluid.prefs.enactor.selfVoicing.parse(elm[0]);
+            that.parseQueue.push(parsedFromElm);
+            that.events.onReadFromDOM.fire(parsedFromElm);
+            that.tts.queueSpeech(fluid.prefs.enactor.selfVoicing.parsedToString(parsedFromElm));
+        }
+    };
+
+    /**
+     * Returns the index of the closest data point from the parseQueue based on the boundary provided.
+     *
+     * @param {Array} parseQueue - An array data points generated from parsing a DOM structure
```

Comment:
This should read something like {ParseQueueElement[]} and the structure defined in a block comment somewhere